### PR TITLE
fix: fail when tests print warnings

### PR DESCRIPTION
**Pull request by shaftoe**
Ref: https://github.com/probot/create-probot-app/issues/309

This is just an idea and it was quick enough to try it out. Probably also `console.error` should be taken into consideration?

### CODE REVIEW

File: templates/basic-ts/test/utils/jest.overrides.js
Comment (gr2m) [inline templates/basic-ts/test/utils/jest.overrides.js:6]: the missing new line should have failed the linting with prettier 🤔

**Review by gr2m — COMMENTED** [review]
It's an interesting idea, but I think I wouldn't introduce a hack to other people's newly setup apps. The deprecation warnings are not a big deal, it's okay to address them when we see them. Once there is a breaking version, we will be forced to address the deprecation messages anyway.

If we find a way to make the CI fail without introducing code to the app templates, then I'm for it. Otherwise I'd say let's leave it. One idea could be to load the logs from the test step and look for "Deprecation". But I'm not sure an action workflow can load its own logs for previous steps while it's still running.

File: templates/basic-ts/test/utils/jest.overrides.js
Comment (shaftoe) [reply templates/basic-ts/test/utils/jest.overrides.js:6]: If I get it right at the moment no linting test step is defined, there's a `lint:fix` inside the Bash script test though: https://github.com/probot/create-probot-app/blob/7c905fabd8fc37141127f687036ab29bfc1d2d27/script/test#L18

**Comment by shaftoe** [comment]
> It's an interesting idea, but I think I wouldn't introduce a hack to other people's newly setup apps. The deprecation warnings are not a big deal, it's okay to address them when we see them. Once there is a breaking version, we will be forced to address the deprecation messages anyway.

Yes, I understand this way is too invasive, fair enough.

> If we find a way to make the CI fail without introducing code to the app templates, then I'm for it. Otherwise I'd say let's leave it. One idea could be to load the logs from the test step and look for "Deprecation". But I'm not sure an action workflow can load its own logs for previous steps while it's still running.

My goal is to shift to CI (or any other automated process) the responsibility to remind us there's actual need to manually update the templates.

At the moment I can see a few ways to achieve that without touching templates as you suggest, it should not be that hard to capture test process output streams and add another downstream test to fail in case some pattern matches are found there.

Before digging into that though I'd like to hear from you if it makes more sense to patch what's there at the moment (`scripts/test` mostly) or drop it and invest some effort and modify/extend `create-probot-app.ts` to be used for testing purposes too.

**Comment by gr2m** [comment]
> modify/extend `create-probot-app.ts` to be used for testing purposes

could you elaborate what your plan would be for that?

**Comment by shaftoe** [comment]
> could you elaborate what your plan would be for that?

I'm trying to understand why tests require Bash in the first place. Node has all the features needed to do all that Bash does (and more, and better IMHO), including forking new processes and inspect their output streams content, interact with `npm`, and so on. I'm not sure why all logic was implemented for Node but not the test script, probably it was the most convenient way to have something working and quickly? I'm sure though there were good reason when that choice was made.

I agree that keeping both Node and Bash in the project is not such a big deal at the moment but I also think keeping two runtimes around doesn't improve the DX much so it's better to avoid unless there's a good reason for that.

If `create-probot-app` as a project is already considered stable, than probably there's no reason to rethink or remake much of it, including the test logic. Otherwise, if it's meant to be more like a _work in progress_ and still open to new features and so on, probably it makes sense to think more about how to test it and investing in a bit of refactoring might make things more consistent and easier to maintain for the future.

For example, how about a new CLI option that just prints the available templates?

```bash
./bin/create-probot-app.js --list-templates
basic-js
basic-ts
checks-js
deploy-js
git-data-js
```

This should be trivial to add and, most importantly, could be used by whatever runtime executes the tests, removing duplication for that particular bit of logic (simple as it might be at the moment) currently shared by both app and tests.

Beside, `script/test` is doing quite a lot at the moment, including dependency install, linting, and so on:

```bash
npm ci
npm run lint:fix
npm run build
[...]
```

and for how I see it, these steps might belong better in the GH Action definition because not strictly related with testing logic (at least I'm positive `lint:fix` shouldn't be there at all).

This also relates partially with what I'm proposing [here](https://github.com/probot/create-probot-app/pull/314#discussion_r553186329), i.e. if we agree to partially rethink `create-probot-app` implementation then we could also keep the _improve testability_ target in mind while doing so.

**Comment by gr2m** [comment]
I think the idea behind using bash for testing is that it allows us to create full end-to-end tests. They are as close as possible to the commands that the user actually runs.

It doesn't mean that all the tests need to be shell scripts. We could leave the shell scripts to test happy paths of the CLI usage, but add integration/unit tests for more granular testing that doesn't require so much overhead?

**Comment by shaftoe** [comment]
> It doesn't mean that all the tests need to be shell scripts. We could leave the shell scripts to test happy paths of the CLI usage, but add integration/unit tests for more granular testing that doesn't require so much overhead?

As usual testing implies some overhead, I can't say how much is too much for the project though. Probably is just easier if I try to apply some of the changes and send a PR after all. For example I'm thinking that we could leverage the matrix strategy here too for testing the various templates

**Comment by gr2m** [comment]
sounds good, do you need anything from me to move forward?

**Comment by shaftoe** [comment]
> sounds good, do you need anything from me to move forward?

With https://github.com/probot/create-probot-app/commit/f0d96d8c5c52ad065daf154de40455d3279de63a I fixed the remaining linting issue and I'm satisfied with the patch, if you're also satisfied than let's go ahead and merge it

**Review by gr2m — APPROVED** [review]
ready to merge once the merge conflicts are addressed 👍🏼

**Comment by gr2m** [comment]
I've invited you to become a member of this repository, so you can merge PRs once the pass all branch requirements 👍🏼

**Comment by shaftoe** [comment]
> I've invited you to become a member of this repository, so you can merge PRs once the pass all branch requirements 👍🏼

Awesome! 🥳

**Pull request merged**
